source/server/listener_manager_impl.cc

```diff
@@ -166,17 +166,19 @@ ListenerImpl::ListenerImpl(const envoy::api::v2::Listener& config, const std::st
   }
 
   bool need_tls_inspector = false;
-  std::set<std::size_t> filter_chains;
+  std::unordered_map<std::size_t, envoy::api::v2::listener::FilterChainMatch> filter_chains;
```

Comment:
This is functionally what I meant. However, I think it would be clearer to do something like:

`std::unordered_set<envoy::api::v2::listener::FilterChainMatch, MessageUtil::hash, Protobuf::util::MessageDifferencer::Equivalent> filter_chains;`

And then just check for it being present, and insert, as the code was doing in the previous version. But if it's a big fight to make the compiler accept the hash and equality functions for unordered_set, I'm fine with leaving it like this.

Reply:
`std::unordered_set<envoy::api::v2::listener::FilterChainMatch, MessageUtil::hash, Protobuf::util::MessageDifferencer::Equivalent>` didn't work, because it requires classes and not methods, but I ended-up pushing them (as operators) to `MessageUtil`.

Let me know what you think.